File: random_forest/add_smos_product_to_points.py
from __future__ import annotations

import argparse
import re
from pathlib import Path

import numpy as np
import pandas as pd
from scipy.spatial import cKDTree

from run_combined_regional_framework import (
    PRODUCT_COL,
    build_combined_training_table,
    load_combined_csv,
)
from utils_data import DEFAULT_CSV, DEFAULT_MAT


LAT_PATH = "/HDFEOS/GRIDS/NpPolarGrid12km/lat"
LON_PATH = "/HDFEOS/GRIDS/NpPolarGrid12km/lon"
DEFAULT_PRODUCT_PATH = "/HDFEOS/GRIDS/NpPolarGrid12km/Data Fields/SI_12km_NH_SNOWDEPTH_5DAY"


def import_h5py():
    try:
        import h5py
    except ImportError as exc:
        raise ImportError("This script needs h5py. Install it with: pip install h5py") from exc
    return h5py


def resolve_lon_360(lon: np.ndarray) -> np.ndarray:
    return np.mod(lon, 360.0)


def latlon_to_unit_xyz(lat: np.ndarray, lon: np.ndarray) -> np.ndarray:
    lon_360 = resolve_lon_360(np.asarray(lon, dtype=float))
    lat_rad = np.radians(np.asarray(lat, dtype=float))
    lon_rad = np.radians(lon_360)
    return np.column_stack(
        [
            np.cos(lat_rad) * np.cos(lon_rad),
            np.cos(lat_rad) * np.sin(lon_rad),
            np.sin(lat_rad),
        ]
    )


def find_first_he5(amsr_dir: Path) -> Path:
    files = sorted(amsr_dir.rglob("*.he5"))
    if not files:
        raise FileNotFoundError(f"No .he5 files found under {amsr_dir}")
    return files[0]


def build_file_index(amsr_dir: Path) -> dict[str, Path]:
    out = {}
    for path in sorted(amsr_dir.rglob("*.he5")):
        match = re.search(r"(20\d{6})", path.name)
        if match and match.group(1) not in out:
            out[match.group(1)] = path
    if not out:
        raise FileNotFoundError(f"No dated .he5 files found under {amsr_dir}")
    return out


def read_grid(sample_file: Path) -> tuple[np.ndarray, np.ndarray]:
    h5py = import_h5py()
    with h5py.File(sample_file, "r") as handle:
        lat = np.asarray(handle[LAT_PATH], dtype=float)
        lon = np.asarray(handle[LON_PATH], dtype=float)
    return lat, lon


def read_product_grid(path: Path, product_path: str, scale: float, invalid_min_cm: float) -> np.ndarray:
    h5py = import_h5py()
    with h5py.File(path, "r") as handle:
        raw = np.asarray(handle[product_path], dtype=float)
    product = raw * scale
    product[(raw >= invalid_min_cm) | (raw < 0)] = np.nan
    return product


def parse_datetime(df: pd.DataFrame) -> pd.Series:
    if "DateTime" in df.columns:
        parsed = pd.to_datetime(df["DateTime"], errors="coerce")
        if parsed.notna().any():
            return parsed
    required = {"Year", "Month", "Day"}
    if not required.issubset(df.columns):
        raise ValueError("Point table needs DateTime or Year/Month/Day columns.")
    return pd.to_datetime(
        pd.DataFrame(
            {
                "year": df["Year"],
                "month": df["Month"],
                "day": df["Day"],
            }
        ),
        errors="coerce",
    )


def load_point_table(args: argparse.Namespace) -> pd.DataFrame:
    if args.combined_csv:
        return load_combined_csv(args.combined_csv)
    if args.input_csv:
        df = pd.read_csv(args.input_csv)
        if "source" not in df.columns:
            df["source"] = "csv"
        return df
    return build_combined_training_table(args.icebridge_mat, args.imb_csv)


def add_product_values(args: argparse.Namespace) -> pd.DataFrame:
    df = load_point_table(args).copy()
    if not {"Latitude", "Longitude"}.issubset(df.columns):
        raise ValueError("Point table needs Latitude and Longitude columns.")

    df["_match_datetime"] = parse_datetime(df)
    df["_match_datestr"] = df["_match_datetime"].dt.strftime("%Y%m%d")

    file_index = build_file_index(args.amsr_dir)
    sample_file = find_first_he5(args.amsr_dir)
    grid_lat, grid_lon = read_grid(sample_file)
    tree = cKDTree(latlon_to_unit_xyz(grid_lat.ravel(), grid_lon.ravel()))

    product_values = np.full(len(df), np.nan)
    distance_km = np.full(len(df), np.nan)
    matched_file = np.full(len(df), "", dtype=object)

    valid_point = df["_match_datestr"].notna() & df["Latitude"].notna() & df["Longitude"].notna()
    for date_str, group in df[valid_point].groupby("_match_datestr"):
        product_file = file_index.get(date_str)
        if product_file is None:
            continue

        product = read_product_grid(
            product_file,
            args.product_path,
            scale=args.product_scale,
            invalid_min_cm=args.invalid_min_raw,
        )
        query = latlon_to_unit_xyz(group["Latitude"].to_numpy(), group["Longitude"].to_numpy())
        chord_dist, flat_idx = tree.query(query, k=1)
        km = 6371.0 * 2.0 * np.arcsin(np.clip(chord_dist / 2.0, 0.0, 1.0))
        values = product.ravel()[flat_idx]

        rows = group.index.to_numpy()
        product_values[rows] = values
        distance_km[rows] = km
        matched_file[rows] = product_file.name

    df[PRODUCT_COL] = product_values
    df["SMOS_Product_Match_Distance_Km"] = distance_km
    df["SMOS_Product_File"] = matched_file
    df = df.drop(columns=["_match_datetime", "_match_datestr"])
    return df


def write_summary(df: pd.DataFrame, out_csv: Path) -> None:
    matched = df[PRODUCT_COL].notna()
    print("Saved:", out_csv)
    print("Rows:", len(df))
    print("Rows with matched SMOS product:", int(matched.sum()))
    if matched.any():
        print(df.loc[matched, PRODUCT_COL].describe())
        if "source" in df.columns:
            print("Matched by source:")
            print(df.loc[matched, "source"].value_counts())


def main() -> None:
    parser = argparse.ArgumentParser(
        description="Add point-matched AMSR/SMOS snow-depth product values to a training/validation point table."
    )
    parser.add_argument("--amsr-dir", type=Path, required=True, help="Local folder containing dated AMSR/AMSR2 .he5 files.")
    parser.add_argument("--out-csv", type=Path, default=Path("data/combined_points_with_smos_product.csv"))
    parser.add_argument("--combined-csv", type=Path, default=None, help="Existing combined point table to augment.")
    parser.add_argument("--input-csv", type=Path, default=None, help="Generic point CSV to augment.")
    parser.add_argument("--icebridge-mat", type=Path, default=DEFAULT_MAT)
    parser.add_argument("--imb-csv", type=Path, default=DEFAULT_CSV)
    parser.add_argument("--product-path", default=DEFAULT_PRODUCT_PATH)
    parser.add_argument(
        "--product-scale",
        type=float,
        default=0.01,
        help="Scale raw product units to meters. AMSR snow-depth product units are cm, so default is 0.01.",
    )
    parser.add_argument(
        "--invalid-min-raw",
        type=float,
        default=110.0,
        help="Raw product values >= this threshold are product flags, not snow depth.",
    )
    args = parser.parse_args()

    args.out_csv.parent.mkdir(parents=True, exist_ok=True)
    df = add_product_values(args)
    df.to_csv(args.out_csv, index=False)
    write_summary(df, args.out_csv)


if __name__ == "__main__":
    main()

File: random_forest/add_smos_product_to_points.m
function add_smos_product_to_points(varargin)
% Add point-matched AMSR/SMOS snow product values to the combined point table.
%
% Example:
%   addpath('random_forest');
%   add_smos_product_to_points( ...
%       'AMSRDir', 'G:\AU_SI12_1-20250103_145027', ...
%       'OutCsv', 'data\combined_points_with_smos_product.csv');
%
% The output CSV adds:
%   SMOS_Product_Snow_Depth_m
%   SMOS_Product_Match_Distance_Km
%   SMOS_Product_File
%
% Product variable:
%   /HDFEOS/GRIDS/NpPolarGrid12km/Data Fields/SI_12km_NH_SNOWDEPTH_5DAY
%
% The product units are cm. Raw values >= 110 are product flags
% (missing, land, open water, multiyear ice, variability, melt), so they are
% written as NaN.

p = inputParser;
addParameter(p, 'AMSRDir', 'G:\AU_SI12_1-20250103_145027', @(x)ischar(x) || isstring(x));
addParameter(p, 'OutCsv', fullfile('data', 'combined_points_with_smos_product.csv'), @(x)ischar(x) || isstring(x));
addParameter(p, 'IceBridgeMat', fullfile('data', 'Master_Dataset_AllFeatures.mat'), @(x)ischar(x) || isstring(x));
addParameter(p, 'ImbCsv', fullfile('data', 'Validation_Dataset_IMB_AllBuoys_Combined_AllData.csv'), @(x)ischar(x) || isstring(x));
addParameter(p, 'InputCsv', '', @(x)ischar(x) || isstring(x));
addParameter(p, 'ProductPath', '/HDFEOS/GRIDS/NpPolarGrid12km/Data Fields/SI_12km_NH_SNOWDEPTH_5DAY', @(x)ischar(x) || isstring(x));
addParameter(p, 'ProductScale', 0.01, @isnumeric);
addParameter(p, 'InvalidMinRaw', 110, @isnumeric);
parse(p, varargin{:});
args = p.Results;

amsrDir = char(args.AMSRDir);
outCsv = char(args.OutCsv);

if strlength(string(args.InputCsv)) > 0
    T = readtable(char(args.InputCsv));
    if ~ismember('source', T.Properties.VariableNames)
        T.source = repmat("csv", height(T), 1);
    end
else
    T = buildCombinedPointTable(char(args.IceBridgeMat), char(args.ImbCsv));
end

required = {'Year','Month','Day','Latitude','Longitude'};
missingRequired = setdiff(required, T.Properties.VariableNames);
if ~isempty(missingRequired)
    error('Point table is missing required columns: %s', strjoin(missingRequired, ', '));
end

he5Files = dir(fullfile(amsrDir, '**', '*.he5'));
if isempty(he5Files)
    error('No .he5 files found under: %s', amsrDir);
end

fileDates = strings(numel(he5Files), 1);
filePaths = strings(numel(he5Files), 1);
for i = 1:numel(he5Files)
    token = regexp(he5Files(i).name, '(20\d{6})', 'match', 'once');
    if ~isempty(token)
        fileDates(i) = string(token);
        filePaths(i) = string(fullfile(he5Files(i).folder, he5Files(i).name));
    end
end
validFiles = fileDates ~= "";
fileDates = fileDates(validFiles);
filePaths = filePaths(validFiles);
if isempty(fileDates)
    error('No dated .he5 files found under: %s', amsrDir);
end
[fileDates, uniqueIdx] = unique(fileDates, 'stable');
filePaths = filePaths(uniqueIdx);

sampleFile = char(filePaths(1));
fprintf('Reading grid from: %s\n', sampleFile);
gridLat = double(h5read(sampleFile, '/HDFEOS/GRIDS/NpPolarGrid12km/lat'));
gridLon = double(h5read(sampleFile, '/HDFEOS/GRIDS/NpPolarGrid12km/lon'));

gridXYZ = latlonToUnitXYZ(gridLat(:), gridLon(:));
fprintf('Building KD-tree with %d grid cells...\n', size(gridXYZ, 1));
tree = KDTreeSearcher(gridXYZ);

pointDate = string(compose('%04d%02d%02d', T.Year, T.Month, T.Day));
productValue = nan(height(T), 1);
matchDistanceKm = nan(height(T), 1);
matchedFile = strings(height(T), 1);

uniquePointDates = unique(pointDate);
for i = 1:numel(uniquePointDates)
    dateStr = uniquePointDates(i);
    fileIdx = find(fileDates == dateStr, 1, 'first');
    if isempty(fileIdx)
        continue;
    end

    rowMask = pointDate == dateStr & isfinite(T.Latitude) & isfinite(T.Longitude);
    if ~any(rowMask)
        continue;
    end

    productFile = char(filePaths(fileIdx));
    fprintf('Matching %s: %d points using %s\n', char(dateStr), sum(rowMask), productFile);
    raw = double(h5read(productFile, char(args.ProductPath)));
    product = raw * args.ProductScale;
    product(raw >= args.InvalidMinRaw | raw < 0) = nan;

    queryXYZ = latlonToUnitXYZ(T.Latitude(rowMask), T.Longitude(rowMask));
    [idx, chordDistance] = knnsearch(tree, queryXYZ);
    km = 6371 * 2 * asin(min(max(chordDistance / 2, 0), 1));

    rowIdx = find(rowMask);
    productValue(rowIdx) = product(idx);
    matchDistanceKm(rowIdx) = km;
    matchedFile(rowIdx) = string(getFileName(productFile));
end

T.SMOS_Product_Snow_Depth_m = productValue;
T.SMOS_Product_Match_Distance_Km = matchDistanceKm;
T.SMOS_Product_File = matchedFile;

outDir = fileparts(outCsv);
if ~isempty(outDir) && ~exist(outDir, 'dir')
    mkdir(outDir);
end
writetable(T, outCsv);

matched = isfinite(T.SMOS_Product_Snow_Depth_m);
fprintf('\nSaved: %s\n', outCsv);
fprintf('Rows: %d\n', height(T));
fprintf('Rows with matched valid product: %d\n', sum(matched));
if any(matched)
    fprintf('Product snow depth summary (m): min=%.4f, median=%.4f, mean=%.4f, max=%.4f\n', ...
        min(T.SMOS_Product_Snow_Depth_m(matched)), ...
        median(T.SMOS_Product_Snow_Depth_m(matched)), ...
        mean(T.SMOS_Product_Snow_Depth_m(matched)), ...
        max(T.SMOS_Product_Snow_Depth_m(matched)));
    if ismember('source', T.Properties.VariableNames)
        disp('Matched rows by source:');
        disp(groupcounts(T(matched, :), 'source'));
    end
end
end

function T = buildCombinedPointTable(iceBridgeMat, imbCsv)
S = load(iceBridgeMat, 'Final_Data');
D = double(S.Final_Data);
if size(D, 2) < 19 && size(D, 1) >= 19
    D = D.';
end
if size(D, 2) < 19
    error('Expected Final_Data to have at least 19 columns, got %dx%d.', size(D, 1), size(D, 2));
end

iceNames = {
    'Year','Month','Day','Latitude','Longitude', ...
    'TB_18V','TB_18H','TB_23V','TB_23H','TB_36V','TB_36H','TB_89V','TB_89H', ...
    'Ice_Thickness_m','Mean_Freeboard_m','Snow_Depth_m', ...
    'Surface_Roughness','KT19_Surface_Temp','MY_Ice_Fraction'};
Ice = array2table(D(:, 1:19), 'VariableNames', iceNames);
Ice.source = repmat("icebridge", height(Ice), 1);
Ice.Hour = nan(height(Ice), 1);
Ice.Distance_Km = nan(height(Ice), 1);

Imb = readtable(imbCsv);
Imb.source = repmat("imb", height(Imb), 1);
if ~ismember('Mean_Freeboard_m', Imb.Properties.VariableNames)
    Imb.Mean_Freeboard_m = nan(height(Imb), 1);
end
if ~ismember('Surface_Roughness', Imb.Properties.VariableNames)
    Imb.Surface_Roughness = nan(height(Imb), 1);
end
if ~ismember('KT19_Surface_Temp', Imb.Properties.VariableNames)
    Imb.KT19_Surface_Temp = nan(height(Imb), 1);
end
if ~ismember('MY_Ice_Fraction', Imb.Properties.VariableNames)
    Imb.MY_Ice_Fraction = nan(height(Imb), 1);
end
if ~ismember('Distance_Km', Imb.Properties.VariableNames)
    Imb.Distance_Km = nan(height(Imb), 1);
end
if ~ismember('Hour', Imb.Properties.VariableNames)
    Imb.Hour = nan(height(Imb), 1);
end

common = {
    'source','Year','Month','Day','Hour','Latitude','Longitude','Distance_Km', ...
    'TB_18V','TB_18H','TB_23V','TB_23H','TB_36V','TB_36H','TB_89V','TB_89H', ...
    'Snow_Depth_m','Ice_Thickness_m','Mean_Freeboard_m','Surface_Roughness', ...
    'KT19_Surface_Temp','MY_Ice_Fraction'};
T = [Ice(:, common); Imb(:, common)];
T = T(isfinite(T.Snow_Depth_m) & T.Snow_Depth_m > 0, :);
end

function xyz = latlonToUnitXYZ(lat, lon)
lon = mod(double(lon), 360);
latRad = deg2rad(double(lat));
lonRad = deg2rad(lon);
xyz = [cos(latRad).*cos(lonRad), cos(latRad).*sin(lonRad), sin(latRad)];
end

function name = getFileName(pathText)
[~, base, ext] = fileparts(pathText);
name = [base ext];
end
